SSPROD-25956: TF support for Organizational Agentless scanning

ravinadhruve10 opened the pull request:
Change summary:
----------------

1. For Agentless scanning, organizational installs will use CF to create stacksets and instances for :-

- IAM role and policy for agentless-scanning,
- KMS primary Key in each account, and each region,
- KMS primary Alias  in each account, and each region,

2. Minor updates to the KMS key module in main.tf
3. Updated the README and added appropriate comments in .tf files for agentless scanning.

Sample main.tf (Agentless scanning, Org Install) :-
```
provider "aws" {
  region = "us-east-1"
}

module "mgmt-account-role" {
  source           = "draios/secure-for-cloud/aws//modules/services/trust-relationship"
  role_name        = "sysdig-secure-2u6g"
  trusted_identity = "arn:aws:iam::0646xxxxxxxx:role/us-east-1-integration01-secure-assume-role"
  external_id      = "81145517xxxxxxxxxxxxxxxxxxxxxx"
}

module "organization-agentless-scanning" {
  source               = "draios/secure-for-cloud/aws//modules/services/agentless-scanning"
  trusted_identity     = "arn:aws:iam::0646xxxxxxxx:role/us-east-1-integration01-secure-assume-role"
  external_id          = "81145517xxxxxxxxxxxxxxxxxxxxxx"
  name                 = "sysdig-secure-scanning-i9yc"
  kms_key_alias        = "test"
  is_organizational    = true
  org_units            = ["r-op65"]
  instrumented_regions = ["us-east-1", "us-west-1", "us-west-2"]
}
```

<!--
Thank you for your contribution!

## Testing your PR

You can pinpoint the pr changes as terraform module source with following format

```
source                    = "github.com/draios/terraform-aws-secure-for-cloud//examples/organizational-ecs?ref=<BRANCH_NAME>"
```


## General recommendations
Check contribution guidelines at https://github.com/draios/terraform-aws-secure-for-cloud/blob/master/CONTRIBUTE.md#contribution-checklist

For a cleaner PR make sure you follow these recommendations:
- Review modified files and delete small changes that were not intended and maybe slip the commit.
- Use Pull Request Drafts for visibility on Work-In-Progress branches and use them on daily mob/pairing for team review
- Unless an external revision is desired, in order to validate or gather some feedback, you are free to merge as long as **validation checks are green-lighted**

## Checklist

- [ ] If `test/fixtures/*/main.tf` files are modified, update:
    - [ ] the snippets in the README.md file under root folder.
    - [ ] the snippets in the README.md file for the corresponding example.
- [ ] If `examples` folder are modified, update:
    - [ ] README.md file with pertinent changes.
    - [ ] `test/fixtures/*/main.tf` in case the snippet needs modifications.
- [ ] If any architectural change has been made, update the diagrams.

-->

ravinadhruve10 committed:
SSPROD-25956: TF support for Organizational Agentless scanning

Change summary:
----------------
For Agentless scanning, organizational installs will use CF to
create stacksets and instances for - scanning role and policy,
KMS primary key, KMS replica key, and KMS Alias.

ravinadhruve10 committed:
SSPROD-25956: Refactoring stackset creation and using mgmt-acc stacksets for KMS resources

Change summary:
----------------
1. Using the approach of using mgmt account stackset and stack instances
   for creating global and non-global resources related to KMS.
2. Cleaner approach to creating resources for organiational installs.
3. Adding instrumented_regions variable to keep track of all additional
   regions to deploy non-global resources to. This will also help get
   rid of the issue of having long TF module snippets for large number
   of regions for org installs.
4. Updated the stackset resources to avoid dependence on single provided
   primary_key variable in the global TF module. Instead now, each child
   account in org installs will refer to the corresponding global primary
   key (id and arn) created within its own account.

ravinadhruve10 commented:
> Based on our discussion, these changes use `deploy_global_resources` in organizational install, to determine if the resource is to be created just once in primary region, or in additional/instrumented regions.
> 
> Running` terraform plan` generated following resources based on the sample main.tf below:- In primary region (us-east-1):
> 

As per the latest refactored changes, running `terraform plan` generated following resources, based on the sample main.tf below.

Sample main.tf :-
```
provider "aws" {
  region = "us-east-1"
}

module "organization-agentless-scanning-global" {
  source                  = "../agentless-scanning"
  trusted_identity        = "arn:aws:iam::0646xxxxxxxx:role/us-east-1-integration01-secure-assume-role"
  external_id             = "81145517xxxxxxxxxxxxxxxxxxxxxx"
  name                    = "sysdig-secure-scanning-i9yc"
  kms_key_alias           = "test"
  deploy_global_resources = true
  is_organizational       = true
  org_units               = ["r-op65"]
}

module "organization-agentless-scanning-non-global" {
  source               = "../agentless-scanning"
  trusted_identity     = "arn:aws:iam::0646xxxxxxxx:role/us-east-1-integration01-secure-assume-role"
  external_id          = "81145517xxxxxxxxxxxxxxxxxxxxxx"
  name                 = "sysdig-secure-scanning-i9yc"
  kms_key_alias        = "test"
  is_organizational    = true
  org_units            = ["r-op65"]
  instrumented_regions = ["us-west-2", "us-west-1"]
}
```

Single globally created resources :-

- IAM Policy
- IAM Role
- Role Attachment
- IAM Role + Policy Stackset for agentless scanning
- IAM Role + Policy Stackset instance (one per account in the org)

In management account, global resources in primary region (us-east-1):

- (KMS Primary key + Alias) Stackset
- (KMS Primary key + Alias) Stackset instance (one in management account)

In management account, non-global resources per instrumented region (us-west-2, us-west-1):

- (KMS Replica key + Alias) Stackset
- (KMS Replica key + Alias) Stackset instance (one per instrumented region in management account)


In each child account, global resources in primary region (us-east-1):

- (KMS Primary key + Alias) Stackset
- (KMS Primary key + Alias) Stackset instance (one per account in the org)

In each child account, non-global resources per instrumented region (us-west-2, us-west-1):

- (KMS Replica key + Alias) Stackset
- (KMS Replica key + Alias) Stackset instance (one per account in the org, per each instrumented region)

ravinadhruve10 committed:
SSPROD-25956: Update stackset resources creation

ravinadhruve10 committed:
SSPROD-25956: Fix terraform fmt error

ravinadhruve10 committed:
SSPROD-25956: Fix syntax issues in Conditions

ravinadhruve10 commented:
Latest status on this PR is described in the JIRA ticket here - https://sysdig.atlassian.net/browse/SSPROD-25956

ravinadhruve10 committed:
feat/SSPROD-25956: Remove KMS replica key and alias creation for Org Installs

ravinadhruve10 commented:
> Latest status on this PR is described in the JIRA ticket here - https://sysdig.atlassian.net/browse/SSPROD-25956

Updated Sample main.tf for reference, as per latest discussions :-

```
provider "aws" {
  region = "us-east-1"
}

module "organization-agentless-scanning" {
  source               = "../agentless-scanning"
  trusted_identity     = "arn:aws:iam::0646xxxxxxxx:role/us-east-1-integration01-secure-assume-role"
  external_id          = "81145517xxxxxxxxxxxxxxxxxxxxxx"
  name                 = "sysdig-secure-scanning-i9yc"
  kms_key_alias        = "test"
  is_organizational    = true
  org_units            = ["r-op65"]
  instrumented_regions = ["us-east-1", "us-west-1", "us-west-2"]
}
```

cgeers committed:
fix(scanning): remove multi region keys

change to remove the use of multi region (replica) keys -- all regions recieve unique primary keys

cgeers reviewed (dismissed):
great job!

cgeers commented:
@ravinadhruve10 I've pushed a small commit to that removes the use of replica keys, and updates documentation. Please give this commit a look and merge at your discretion.

ravinadhruve10 commented on modules/services/agentless-scanning/README.md:
```diff
@@ -29,12 +29,19 @@ No modules.
+| [aws_cloudformation_stack_set.mgmt_acc_resources_stackset](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cloudformation_stack_set) | resource |
+| [aws_cloudformation_stack_set.ou_resources_stackset](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cloudformation_stack_set) | resource |
+| [aws_cloudformation_stack_set.scanning_role_stackset](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cloudformation_stack_set) | resource |
+| [aws_cloudformation_stack_set_instance.mgmt_acc_global_stackset_instance](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cloudformation_stack_set_instance) | resource |
```
These 4 stackset_instances (`mgmt_acc_global_stackset_instance`, `mgmt_acc_nonglobal_stackset_instance`, `ou_global_stackset_instance`, `ou_nonglobal_stackset_instance`) should be consolidated into 2 actually - `mgmt_acc_stackset_instance`, `ou_global_stackset_instance`

ravinadhruve10 replied:
Fixed.

ravinadhruve10 commented:
> great job!

Thank you! :)

I have a minor change in this PR, will push that along with the README.md fix.

ravinadhruve10 committed:
Update README.md

Signed-off-by: Ravina Dhruve <136399755+ravinadhruve10@users.noreply.github.com>

ravinadhruve10 committed:
Remove multiregion property in stacksets for KMS Primary Key

Since we are getting rid of KMS Replicas, we don't need to create multiregion based KMS primary keys.

Signed-off-by: Ravina Dhruve <136399755+ravinadhruve10@users.noreply.github.com>

cgeers reviewed (dismissed):
nice catch!

ravinadhruve10 committed:
Update comments in main.tf

Signed-off-by: Ravina Dhruve <136399755+ravinadhruve10@users.noreply.github.com>
